scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala

```diff
@@ -118,261 +79,396 @@ class EscapeHatch(
+        case f @ EscapeFilter(_, _, _, Some(end))
+            if f.matches(ruleName) && end.offset >= position =>
+          true
+        case _ => false
+      }
+      maybeFilter match {
+        case Some(filter) => (false, Some(filter))
+        case None => (true, None)
+      }
+    }
 
-  private sealed abstract class Escape
-  private object Escape {
-    case class UntilEOF(
-        anchorPosition: AnchorPosition,
-        rules: String,
-        toogle: Toogle
-    ) extends Escape
-
-    case class Expression(
-        expressionPosition: ExpressionPosition,
-        anchorPosition: AnchorPosition,
-        rules: String
-    ) extends Escape
-
-    case class ExpressionLine(
-        expressionPosition: ExpressionPosition,
-        anchorPosition: AnchorPosition,
-        rules: String
-    ) extends Escape
+    def disableEscapes: Iterable[EscapeFilter] = escapeTree.values.flatten
   }
 
-  private sealed abstract class Toogle
-  private object Toogle {
-    case object Disable extends Toogle
-    case object Enable extends Toogle
-  }
+  private object AnnotatedEscapes {
+    private val SuppressWarnings = "SuppressWarnings"
+    private val SuppressAll = "all"
+    val OptionalRulePrefix = "scalafix:"
 
-  private val unusedScalafixSupression = RuleName("UnusedScalafixSupression")
+    def apply(tree: Tree): AnnotatedEscapes = {
+      val builder = TreeMap.newBuilder[EscapeOffset, List[EscapeFilter]]
 
-  private val unusedEnableWarning =
-    LintCategory
-      .warning(
-        "Enable",
-        "This comment would enable a rule that was not disabled (eg: typo in the rules)"
-      )
-      .withOwner(unusedScalafixSupression)
-
-  private val FilterDisable = "\\s?scalafix:off\\s?(.*)".r
-  private val FilterEnable = "\\s?scalafix:on\\s?(.*)".r
-  private val FilterExpression = "\\s?scalafix:ok\\s?(.*)".r
-
-  private def splitRules(rules: String): Set[String] =
-    rules.split("\\s+").toSet
-
-  def apply(tree: Tree, associatedComments: AssociatedComments): EscapeHatch = {
-    val escapes = List.newBuilder[Escape]
-    val unusedEnable = List.newBuilder[LintMessage]
-    val visitedFilterExpression = mutable.Set.empty[AnchorPosition]
-
-    var currentlyDisabledRules = Set.empty[String]
-    def trackUnusedRules(
-        rules: String,
-        toogle: Toogle,
-        anchorPosition: AnchorPosition): Unit = {
-      toogle match {
-        case Toogle.Disable =>
-          currentlyDisabledRules = currentlyDisabledRules ++ splitRules(rules)
-
-        case Toogle.Enable =>
-          val enabledRules = splitRules(rules)
-          val enabledNotDisabled = enabledRules -- currentlyDisabledRules
-
-          enabledNotDisabled.foreach(
-            rule => unusedEnable += unusedEnableWarning.at(anchorPosition.value)
-          )
+      def hasSuppressWarnings(mods: List[Mod]): Boolean =
+        mods.exists {
+          case Mod.Annot(Init(Type.Name(SuppressWarnings), _, _)) => true
```

Comment:
What happens for `@java.lang.SuppressWarnings`?  I agree it makes sense to do this syntactically in order to enable syntactic rules like DisableSyntax to support annotations. Can we add an entry in the docs explaining the  detection of `@SuppressWarnings`?

> The `@SuppressWarnings` annotation is detected without compilation. Any annotation matching the syntax `@SuppressWarnings(..)` regardless if it's `java.lang.SupressWarnings` or not will trigger the suppression mechanism. This is done in order to support `@SuppressWarnings` for syntactic rules like `DisableSyntax`.


To support qualifying the annotation, I recommend we do the following:

```scala
case Mod.Annot(Init(Type.Name(SuppressWarnings), _, List(List(_)))) => true
case Mod.Annot(Init(Type.Select(_, Type.Name(SuppressWarnings)), _, List(List(_)))) => true // `@java.lang.SuppressWarnings(..)`
```

Reply:
>What happens for @java.lang.SuppressWarnings?

Nice catch! Fixed

>Can we add an entry in the docs explaining the detection of @SuppressWarnings?

Done